### PR TITLE
rstan::optimizing backend implementation

### PR DESCRIPTION
**Pull request by Dashadower**
Looks like a  dummy `SBC_fit_to_diagnostics` function is needed.

### CODE REVIEW

**Comment by hyunjimoon** [comment]
I faced the following message.
```
> result_binom_opt <- compute_results(datasets_binom, backend_binom_opt, thin_ranks = 1)
Dataset 1 resulted in error when post-processing the fit.
Calling `recompute_statistics` after you've found and fixed the problem could let you move further without refitting
Error: $ operator not defined for this S4 class


Dataset 2 resulted in error when post-processing the fit.
Calling `recompute_statistics` after you've found and fixed the problem could let you move further without refitting
Error: $ operator not defined for this S4 class
```
This is the reproducible code
```
generator_binom <- function(lambda_mu, lambda_sigma, fixed_sim_args){
  # fixed value across simulated datasets
  # experiment settings
  nobs <- fixed_sim_args$nobs
  link_type <- fixed_sim_args$link_type
  # clamped parameters

  # generate
  eta <- rnorm(1, mean = lambda_mu, sd=lambda_sigma)
  mu <- invtf_param_vec(eta, link_type = link_type)
  Y <- rbinom(nobs, size = nsize, prob = mu) 
  list(
    parameters = list(eta = eta),
    generated = list(nobs= nobs, 
                     lambda_mu = lambda_mu, lambda_log_sigma = log(lambda_sigma), 
                     Y = Y)
  )
}
fixed_sim_args_binom <- list(nobs = nobs, link_type = 1)

# initial prior hyperparameters
lambda_mu <- 1
lambda_sigma <- 10

datasets_binom <- generate_datasets(SBC_generator_function(generator_binom, lambda_mu, lambda_sigma, fixed_sim_args_binom), n_datasets = nsims)
rstan_binom_mod <- stan_model("models/binom-laplace.stan")
backend_binom_opt <- SBC_backend_rstan_optimizing(rstan_binom_mod, draws = ndraws)
result_binom_opt <- compute_results(datasets_binom, backend_binom_opt, thin_ranks = 1)
```

**Comment by hyunjimoon** [comment]
I faced the following message. It turned out `datasets_binom` was wrong but does this mean that error handling could be improved?
```
> result_binom_opt <- compute_results(datasets_binom, backend_binom_opt, thin_ranks = 1)
Dataset 1 resulted in error when post-processing the fit.
Calling `recompute_statistics` after you've found and fixed the problem could let you move further without refitting
Error: $ operator not defined for this S4 class


Dataset 2 resulted in error when post-processing the fit.
Calling `recompute_statistics` after you've found and fixed the problem could let you move further without refitting
Error: $ operator not defined for this S4 class
```
This is the reproducible code
```
generator_binom <- function(lambda_mu, lambda_sigma, fixed_sim_args){
  # fixed value across simulated datasets
  # experiment settings
  nobs <- fixed_sim_args$nobs
  link_type <- fixed_sim_args$link_type
  # clamped parameters

  # generate
  eta <- rnorm(1, mean = lambda_mu, sd=lambda_sigma)
  mu <- invtf_param_vec(eta, link_type = link_type)
  Y <- rbinom(nobs, size = nsize, prob = mu) 
  list(
    parameters = list(eta = eta),
    generated = list(nobs= nobs, 
                     lambda_mu = lambda_mu, lambda_log_sigma = log(lambda_sigma), 
                     Y = Y)
  )
}
fixed_sim_args_binom <- list(nobs = nobs, link_type = 1)

# initial prior hyperparameters
lambda_mu <- 1
lambda_sigma <- 10

datasets_binom <- generate_datasets(SBC_generator_function(generator_binom, lambda_mu, lambda_sigma, fixed_sim_args_binom), n_datasets = nsims)
rstan_binom_mod <- stan_model("models/binom-laplace.stan")
backend_binom_opt <- SBC_backend_rstan_optimizing(rstan_binom_mod, draws = ndraws)
result_binom_opt <- compute_results(datasets_binom, backend_binom_opt, thin_ranks = 1)
```

**Comment by hyunjimoon** [comment]
"Error: Don't know how to transform an object of class 'NULL' to any supported draws format."
```
...
Iteration 0 - lambda loss: 1.042505 eta loss: 0.185725
Iteration 1 - lambda loss: 1.042505 eta loss: 0.185515
 - 23 (29%) fits had at least one Rhat > 1.01. Largest Rhat was 1.089.
 - 7 (9%) fits had tail ESS undefined or less than half of the maximum rank, potentially skewing 
the rank statistics. The lowest tail ESS was 25.
...
Dataset 4 resulted in error when post-processing the fit.
Calling `recompute_statistics` after you've found and fixed the problem could let you move further without refitting
Error: Don't know how to transform an object of class 'NULL' to any supported draws format.


Dataset 5 resulted in error when post-processing the fit.
Calling `recompute_statistics` after you've found and fixed the problem could let you move further without refitting
Error: Don't know how to transform an object of class 'NULL' to any supported draws format.


Too many datasets produced errors. Further error messages not shown.

 - 80 (100%) fits resulted in an error.
Not all diagnostics are OK.
You can learn more by inspecting $default_diagnostics, $backend_diagnostics 
and/or investigating $outputs/$messages/$warnings for detailed output from the backend.
Error: Don't know how to transform an object of class 'NULL' to any supported draws format.
In addition: There were 50 or more warnings (use warnings() to see the first 50)
```

File: R/backends.R [hunk 1]
```diff
@@ -151,6 +151,62 @@ SBC_backend_hash_for_cache.SBC_backend_rstan_sample <- function(backend) {
...
+#' @param model a `stanmodel` object (created via `rstan::stan_model`)
+#' @param ... other arguments passed to `optimizing` (number of iterations, ...).
+#'   Argument `data` cannot be set this way as they need to be
+#'   controlled by the package.
+#' @export
+SBC_backend_rstan_optimizing <- function(model, ...) {
+  stopifnot(inherits(model, "stanmodel"))
+  message("Please note point estimate methods do not return any meaningful MCMC diagnostic values.")
+  args <- list(...)
+  unacceptable_params <- c("data")
```
Comment (martinmodrak) [inline R/backends.R:166]: Shouldn't we also have `hessian` in there?

File: R/backends.R [hunk 2]
```diff
@@ -151,6 +151,62 @@ SBC_backend_hash_for_cache.SBC_backend_rstan_sample <- function(backend) {
...
+  if(any(names(args) %in% unacceptable_params)) {
+    stop(paste0("Parameters ", paste0("'", unacceptable_params, "'", collapse = ", "),
+                " cannot be provided when defining a backend as they need to be set ",
+                "by the SBC package"))
+  }
+
+  args$hessian = TRUE
```
Comment (martinmodrak) [inline R/backends.R:173]: We probably also need to change the default of the `draws` arg (as it is otherwise zero)

File: R/backends.R [hunk 3]
```diff
@@ -151,6 +151,62 @@ SBC_backend_hash_for_cache.SBC_backend_rstan_sample <- function(backend) {
...
+
+
+#' @export
+SBC_fit.SBC_backend_rstan_optimizing <- function(backend, generated, cores) {
+  structure(list(fit_list=do.call(rstan::optimizing,
+                combine_args(list(object = backend$model,
+                                  data = generated),
+                             backend$args)
+                ))
+            , class = "RStanOptimizingFit")
```
Comment (martinmodrak) [inline R/backends.R:185]: Backends usually don't return a copy of data within the fit, so I would avoid that. In fact I think just adding a custom class to the result of `optimizing` should be sufficient (weird that `rstan` doesn't give the return value a class by itself).

File: R/backends.R [hunk 4]
```diff
@@ -151,6 +151,62 @@ SBC_backend_hash_for_cache.SBC_backend_rstan_sample <- function(backend) {
...
+  res <- data.frame(
+    max_chain_time = 0,
+    n_failed_chains = 0,
+    n_divergent = 0,
+    n_max_treedepth =  0,
+    n_rejects = 0
+    #
+  )
+  class(res) <- c("SBC_nuts_diagnostics", class(res))
+  res
```
Comment (martinmodrak) [inline R/backends.R:204]: That looks wrong to me - there is no reason optimizing fit should have the `SBC_nuts_diagnostics` class. Similarly to what is done in the `implementing_backends` notebook, I think we should either: a) don't implement the `SBC_fit_to_diagnostics` method at all (if we don't have any diagnostic) or b) declare a new class that will contain the optimizing-specific diagnostics (I guess we should at least check the `return_code` field). A simple example where the optimizer does not converge:

```
m <- stan_model(model_code = 'parameters {real y;} model {target += y;}')
f <- optimizing(m, hessian = TRUE)
```

**Review by martinmodrak — COMMENTED** [review]
Hi, thanks for working on the backend - I have a bunch of technical issues (commented directly on code). 

Additionally, I think there should be an implementation of `SBC_backend_iid_samples` (as the samples are independent unless you do importance resampling - and if you do importance resampling, thinning won't help anyway). Additionally an implementation of `SBC_backend_hash_for_cache` is needed (could exactly mirror `SBC_backend_hash_for_cache.SBC_backend_rstan_sample`). If the description in the `implementing_backends` vignette is unclear on any point, feel free to comment (if you are having trouble following the recommendations there, it is definitely a sign that the vignette needs impropving)

Finally, could you prepare a short vignette demonstrating the backend (ideally, it would show a simple model where `optimizing` actually passes SBC and another one where it fails). But I can definitely write the vignette myself as well, if you don't have time.

Thanks again!

**Comment by hyunjimoon** [comment]
Thanks for this @martinmodrak! I will attend to this by the end of this year hopefully with my exploration of self-calibration.

**Pull request merged**